au4/src/au3wrap/internal/processinginteraction.cpp

```diff
@@ -61,6 +67,116 @@ bool ProcessingInteraction::changeClipTitle(const processing::ClipKey& clipKey,
     return true;
 }
 
+void ProcessingInteraction::clearClipboard()
+{
+    s_clipboard.data.clear();
+}
+
+bool ProcessingInteraction::paste(double begin, processing::TrackId trackId)
+{
+    WaveTrack* dstWaveTrack = DomAccessor::findWaveTrack(projectRef(), TrackId(trackId));
+    IF_ASSERT_FAILED(dstWaveTrack) {
+        return false;
+    }
+
+    if (s_clipboard.data.empty()) {
+        return false;
+    }
+
+    // check if every clip fits into track first
+    for (const auto& data : s_clipboard.data) {
+        WaveTrack* origWaveTrack = DomAccessor::findWaveTrack(projectRef(), TrackId(data.track->GetId()));
+        IF_ASSERT_FAILED(dstWaveTrack) {
+            return false;
+        }
+
+        double insertDuration = 0;
+        if (data.clipKey.index == muse::nidx) {
+            std::shared_ptr<WaveClip> leftClip = origWaveTrack->GetLeftmostClip();
+            std::shared_ptr<WaveClip> rightClip = origWaveTrack->GetRightmostClip();
+            insertDuration = rightClip->End() - leftClip->Start();
+        } else {
+            std::shared_ptr<WaveClip> clip = DomAccessor::findWaveClip(origWaveTrack, data.clipKey.index);
+            IF_ASSERT_FAILED(clip) {
+                return false;
+            }
+
+            insertDuration = clip->End() - clip->Start();
+        }
+
+        // throws incosistency exception if project tempo is not set, see:
+        // au4/src/au3wrap/internal/au3project.cpp: 92
+        // Paste func throws exception if there's not enough space to paste in (exception handler calls BasicUI logic)
+        if (!dstWaveTrack->IsEmpty(begin, begin + insertDuration)) {
+            LOGDA() << "not enough space to paste clip into";
+            //! TODO AU4: show dialog
+            return false;
+        }
+    }
+
+    // paste
+    for (const auto& data : s_clipboard.data) {
+        dstWaveTrack->Paste(begin, *data.track);
+        int numClips = dstWaveTrack->GetNumClips();
+
+        processing::ProcessingProjectPtr prj = globalContext()->currentProcessingProject();
+        for (int i = 0; i < numClips; ++i) {
+            std::shared_ptr<WaveClip> clip = DomAccessor::findWaveClip(dstWaveTrack, i);
+            prj->onClipAdded(DomConverter::clip(dstWaveTrack, clip.get(), i));
```

Comment:
I think we should only send clips that were added, not all clips from the current track

Reply:
Fixed